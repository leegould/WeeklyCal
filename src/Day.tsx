import React, {Component} from 'react';
import {StyleSheet, Text, View, FlatList} from 'react-native';
import { CalendarDay } from './shared';

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: 'beige',
        padding: 5,
        margin: 5,
        flexDirection: 'column',
        justifyContent: 'flex-start',
    },
    row: {
        flexDirection: 'row',
        justifyContent: 'space-between',
        borderColor: 'grey',
        borderBottomWidth: 1,
    }
});

type Props = {
    day: CalendarDay,
    isToday?: boolean,
};

export default class Day extends Component<Props> {
    shouldComponentUpdate(nextProps: Props) {
        console.log('Day.shouldComponentUpdate', nextProps, nextProps.day.events.length, this.props.day.events.length, nextProps.day.events.length !== this.props.day.events.length);
        // if (nextProps.day.date !== this.props.day.date || nextProps.day.events.length !== this.props.day.events.length) {
        //     return true;
        // }
        // return false;
        return true;
    }

    render() {
        console.log('Day.render', this.props.day);
        return(
            <View style={styles.container}>
                <View style={styles.row}>
                    <Text>{this.props.day.date.format('dddd')}</Text>
                    <Text>{this.props.day.date.format('DD')}</Text>
                </View>
                {this.props.day.events && 
                    <FlatList
                        data={this.props.day.events}
                        keyExtractor={(item, index) => `event_key_${index}`}
                        renderItem={({item}) => {
                            console.log('item', item);
                            return(<Text>{item.title}</Text>);
                        }}
                    />
                }
            </View>
        );
    }
}

