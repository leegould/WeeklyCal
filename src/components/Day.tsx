import React, {Component} from 'react';
import { StyleSheet, View, Text, FlatList, Animated, TouchableHighlight, TouchableOpacity } from 'react-native';
import { Icon } from 'react-native-elements'
import moment from 'moment';
import { Day, OptionsState, CalendarEvent } from '../types';
import { CalendarEventReadable } from 'react-native-calendar-events';
import InlineEvent from './InlineEvent';

type Props = {
    isFetching: boolean,
    day: Day,
    isToday?: boolean,
    options: OptionsState,
    navigation: {
        navigate: Function,
    },
    onExpand: Function,
    expanded?: boolean,
    onAddEvent: Function,
};

type State = {
    fade: Animated.Value,
};

export default class DayEvents extends Component<Props, State> {
    constructor(props: Props) {
        super(props);
        this.state = {
            fade: new Animated.Value(1),
        }
    }

    componentDidUpdate() {
        if (this.props.isFetching) {
            this.fadeOut();
        } else {
            this.fadeIn();
        }
    }

    fadeIn() {
        this.state.fade.setValue(0);
        Animated.timing(                  
            this.state.fade,            
            {
                toValue: 1,                   
                duration: 500,              
            }
        ).start();
    }

    fadeOut() {
        this.state.fade.setValue(1);
        Animated.timing(                  
            this.state.fade,            
            {
                toValue: 0,                   
                duration: 750,              
            }
        ).start();
    }

    onAddClick() {
        this.props.navigation.navigate('Event', { date: moment(this.props.day.date) });
    }

    onEditClick(item: CalendarEventReadable) {
        const { startDate } = item;
        this.props.navigation.navigate('Event', { date: moment(startDate), event: item })
    }

    onAddEvent(event: CalendarEvent) {
        this.props.onAddEvent(event);
    }

    render() {
        // console.log('Day.render', this.props.day, this.state.events);
        const { day, options, expanded } = this.props;

        return(
            <Animated.View style={[styles.container, {opacity: this.state.fade}]}>
                <TouchableHighlight onPress={() => options.dayAddLink && this.onAddClick()} underlayColor={'beige'} hitSlop={{top: 10, left: 10, bottom: 10, right: 10}}>
                    <View style={styles.row}>
                        {options.dayAddLink &&
                        <Icon name='calendar-plus' type='material-community' color='green' onPress={() => this.onAddClick()} size={expanded ? 24 : 18} />
                        }
                        <Animated.Text style={[styles.header, { fontSize: expanded ? 20 : 14 , opacity: this.state.fade}]}>{`${moment(day.date).format('ddd')} ${moment(day.date).format('DD')}`}</Animated.Text>
                        <Icon name={`arrow-${expanded ? 'collapse' : 'expand'}`} type='material-community' color='green' onPress={() => this.props.onExpand(day)} size={expanded ? 24 : 18} />
                    </View>
                </TouchableHighlight>
                {day.events && 
                    <FlatList
                        data={day.events}
                        keyExtractor={(item, index) => `event_key_${index}`}
                        renderItem={({item}) => {
                            // console.log('Day.item', item);
                            return(
                                <TouchableOpacity onPress={
                                    () => this.onEditClick(item)
                                }>
                                    <Animated.View style={[styles.eventContainer, {opacity: this.state.fade}]}>
                                        {options.eventColor &&
                                        <View style={[styles.eventCalendar, { backgroundColor: item.calendar ? item.calendar.color : 'transparent' }]} />
                                        }
                                        <Text style={[styles.eventTitle, { fontSize: expanded ? 18 : 14 }]}>
                                            {item.title}
                                        </Text>
                                    </Animated.View>
                                </TouchableOpacity>
                            );
                        }}
                    />
                }
                <InlineEvent date={day.date} onAddEvent={(event: CalendarEvent) => this.onAddEvent(event)} />
            </Animated.View>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: 'beige',
        padding: 5,
        margin: 5,
        borderRadius: 5,
        flexDirection: 'column',
        justifyContent: 'flex-start',
    },
    row: {
        flexDirection: 'row',
        justifyContent: 'space-between',
        borderColor: 'grey',
        borderBottomWidth: 0.5,
        marginBottom: 5,
    },
    header: {
        color: 'gray',
    },
    eventContainer: {
        flexDirection: 'row',
        justifyContent: 'flex-start',
        alignItems: 'center',
        marginBottom: 2,
        padding: 2,
    },
    eventCalendar: {
        width: 5,
        height: 5,
        borderRadius: 2,
        marginRight: 3,
    },
    eventTitle: {
        color: 'gray',
    },
});
