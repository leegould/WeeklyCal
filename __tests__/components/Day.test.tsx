import React from 'react';
import moment from 'moment';
import renderer from 'react-test-renderer';
import Day from '../../src/components/Day';

jest.useFakeTimers();

test('renders correctly with expected args', () => {
    const props = {
        navigation: {
            navigate: () => {},
        },
        onExpand: () => {},
        onAddEvent: () => {},
        date: moment('25/12/2005', 'DD/MM/YYYY'),
        isFetching: false,
        day: {
            date: moment('25/12/2005', 'DD/MM/YYYY').toDate(),
            events: [],
        },
        isToday: false,
        options: {
            resetDate: false,
            eventColor: false,
            dayAddLink: false,
            rollingWeek: false,
            inlineAdd: false,
            eventRowBorder: false,
        },
        expanded: false,
    }

    const tree = renderer.create(
        <Day {...props} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
});
