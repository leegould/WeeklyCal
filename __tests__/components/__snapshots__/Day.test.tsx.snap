// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with expected args 1`] = `
<View
  style={
    Object {
      "backgroundColor": "beige",
      "borderRadius": 5,
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "flex-start",
      "margin": 5,
      "marginBottom": 2,
      "opacity": 1,
      "padding": 5,
    }
  }
>
  <View
    accessible={true}
    hitSlop={
      Object {
        "bottom": 10,
        "left": 10,
        "right": 10,
        "top": 10,
      }
    }
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={null}
  >
    <View
      style={
        Object {
          "borderBottomWidth": 0.5,
          "borderColor": "#C2272D",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 3,
          "paddingBottom": 2,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "gray",
            "fontSize": 14,
            "opacity": 1,
          }
        }
      >
        Sun 25
      </Text>
      <View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "justifyContent": "center",
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "green",
                  "fontSize": 18,
                },
                Object {
                  "backgroundColor": "transparent",
                },
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            testID="iconIcon"
          >
            ï˜–
          </Text>
        </View>
      </View>
    </View>
  </View>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</View>
`;
